arangod/Aql/ModificationExecutor2.cpp

```diff
@@ -0,0 +1,286 @@
+// * if value is anything else, we return an error.
+Result ModificationExecutorHelpers::getKeyAndRevision(CollectionNameResolver const& resolver,
+                                                      AqlValue const& value,
+                                                      std::string& key, std::string& rev,
+                                                      bool ignoreRevision) {
+  TRI_ASSERT(key.empty());
+  TRI_ASSERT(rev.empty());
+  if (value.isObject()) {
+    bool mustDestroy;
+    AqlValue sub = value.get(resolver, StaticStrings::KeyString, mustDestroy, false);
+    AqlValueGuard guard(sub, mustDestroy);
+
+    if (sub.isString()) {
+      key.assign(sub.slice().copyString());
+
+      if (!ignoreRevision) {
+        bool mustDestroyToo;
+        AqlValue subTwo =
+            value.get(resolver, StaticStrings::RevString, mustDestroyToo, false);
+        AqlValueGuard guard(subTwo, mustDestroyToo);
+        if (subTwo.isString()) {
+          rev.assign(subTwo.slice().copyString());
+        } else {
+          return Result(TRI_ERROR_ARANGO_DOCUMENT_TYPE_INVALID,
+                        std::string{"Expected _rev as string, but got "} +
+                            value.slice().typeName());
+        }
+      }
+    } else {
+      return Result(TRI_ERROR_ARANGO_DOCUMENT_TYPE_INVALID,
+                    std::string{"Expected _key as string, but got "} +
+                        value.slice().typeName());
+    }
+  } else if (value.isString()) {
+    key.assign(value.slice().copyString());
+    rev.clear();
+  } else {
+    return Result(TRI_ERROR_ARANGO_DOCUMENT_KEY_MISSING,
+                  std::string{"Expected object or string, but got "} +
+                      value.slice().typeName());
+  }
+  return Result{};
+}
+
+// Builds an object "{ _key: key, _rev: rev }" if rev is nonempty and ignoreRevs is false, and
+// "{ _key: key, _rev: null }" otherwise.
+Result ModificationExecutorHelpers::buildKeyDocument(VPackBuilder& builder,
+                                                     std::string const& key,
+                                                     std::string const& rev,
+                                                     bool ignoreRevision) {
+  builder.openObject();
+  builder.add(StaticStrings::KeyString, VPackValue(key));
+  if (ignoreRevision && !rev.empty()) {
+    builder.add(StaticStrings::RevString, VPackValue(rev));
+  } else {
+    builder.add(StaticStrings::RevString, VPackValue(VPackValueType::Null));
+  }
+  builder.close();
+  return Result{};
+}
+
+template <typename FetcherType, typename ModifierType>
+ModificationExecutor2<FetcherType, ModifierType>::ModificationExecutor2(FetcherType& fetcher,
+                                                                        Infos& infos)
+    : _infos(infos), _fetcher(fetcher), _modifier(infos) {
+  // important for mmfiles
+  // TODO: Why is this important for mmfiles?
```

Comment:
In MMFiles we need to make sure that the data is not moved in memory or collected for this collection as soon as we start writing to it.
This pin makes sure that all of the memory moving is not executed and all Pointers we get from a collection stay correct until we release this pin